cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowProgramRunner.java

```diff
@@ -101,25 +115,111 @@ public ProgramController run(Program program, ProgramOptions options) {
     WorkflowSpecification workflowSpec = appSpec.getWorkflows().get(program.getName());
     Preconditions.checkNotNull(workflowSpec, "Missing WorkflowSpecification for %s", program.getName());
 
-    RunId runId = ProgramRunners.getRunId(options);
+    final RunId runId = ProgramRunners.getRunId(options);
 
     // Setup dataset framework context, if required
     if (datasetFramework instanceof ProgramContextAware) {
       ProgramId programId = program.getId();
       ((ProgramContextAware) datasetFramework).initContext(programId.run(runId));
     }
 
+    // List of all Closeable resources that needs to be cleanup
+    final List<Closeable> closeables = new ArrayList<>();
+    try {
+      PluginInstantiator pluginInstantiator = createPluginInstantiator(options, program.getClassLoader());
+      if (pluginInstantiator != null) {
+        closeables.add(pluginInstantiator);
+      }
+
+      WorkflowDriver driver = new WorkflowDriver(program, options, hostname, workflowSpec, programRunnerFactory,
+                                                 metricsCollectionService, datasetFramework, discoveryServiceClient,
+                                                 txClient, runtimeStore, cConf, pluginInstantiator,
+                                                 secureStore, secureStoreManager, messagingService);
+
+      // Controller needs to be created before starting the driver so that the state change of the driver
+      // service can be fully captured by the controller.
+      final ProgramController controller = new WorkflowProgramController(program, driver, serviceAnnouncer, runId);
+      final String twillRunId = options.getArguments().getOption(ProgramOptionConstants.TWILL_RUN_ID);
+      controller.addListener(new AbstractListener() {
+        @Override
+        public void init(ProgramController.State state, @Nullable Throwable cause) {
+          // Get start time from RunId
+          long startTimeInSeconds = RunIds.getTime(controller.getRunId(), TimeUnit.SECONDS);
+          if (startTimeInSeconds == -1) {
+            // If RunId is not time-based, use current time as start time
+            startTimeInSeconds = TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis());
+          }
+          runtimeStore.setStart(program.getId(), runId.getId(), startTimeInSeconds, twillRunId,
+                                options.getUserArguments().asMap(), options.getArguments().asMap());
+
+          switch (state) {
```

Comment:
actually no need to have KILLED state here, since when actions such as SUSPEND/KILLED would happen init listener would have executed at that time. This is similar to the https://github.com/caskdata/cdap/blob/develop/cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java#L341